File: visualPatterns/pyramid/Pyramid.java
import java.util.Arrays;

public class Pyramid { 

    static String cellPattern = "*";
    static String emptyPattern = " ";
    public static void main(String[] args) {
        createPyramidWithNRows(5);
        createInvertedPyramidWithNRows(5);
    }

    static void createPyramidWithNRows(Integer numOfRows) {
        Integer totalSpace = numOfRows + numOfRows - 1;
        Integer midIndex = (int)(Math.ceil(totalSpace/2));
        String[] row = new String[totalSpace];

        Arrays.fill(row, emptyPattern);

        for(int i = 0; i< numOfRows; i++) {
            row[midIndex + i] = cellPattern;
            row[midIndex - i] = cellPattern;
            System.out.println(String.join("", row));
        }
    }

    static void createInvertedPyramidWithNRows(Integer numOfRows) {
        Integer totalSpace = numOfRows + numOfRows - 1;
        Integer midIndex = (int)(Math.ceil(totalSpace/2));
        String[] row = new String[totalSpace];

        Arrays.fill(row, cellPattern);

        for(int i = numOfRows - 1; i>=0; i--) {
            row[midIndex + i] = emptyPattern;
            row[midIndex - i] = emptyPattern;
            System.out.println(String.join("", row));
        }
    }
}

File: Pyramid.ts
const cellPattern = "*";
const emptyPattern = " ";

const createPyramidWithNRows = (numOfRows: number) => {
   const totalSpace = numOfRows + numOfRows - 1;
   const row = new Array(totalSpace).fill(emptyPattern);
   const mid = Math.ceil(totalSpace / 2) - 1;
   console.log("mid: ", mid);
   for (let i = 0; i < numOfRows; i++) {
       row[mid + i] = cellPattern;
       row[mid - i] = cellPattern;
       console.log(row.join(""));
   }
};


const createInvertedPyramidOfNRows = (numOfRows: number) => {
   const totalSpace = numOfRows + numOfRows - 1;
   const row = new Array(totalSpace).fill(cellPattern);
   const mid = Math.ceil(totalSpace / 2) - 1;

   for (let i = 0; i <= mid; i++) {
      console.log(row.join(""));
      row[i] = emptyPattern;
      row[totalSpace - 1 - i] = emptyPattern; 
   }
};